File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.py
import sys
def calculate_power(n,m,board):
    visited=[[False]*n for _ in range(m)]
    dy=[-1,1,0,0]
    dx=[0,0,-1,1]
    t_white=t_blue=0
    def dfs(y,x,color):
        cnt=1
        visited[y][x]=True
        for i in range(4):
            ny=y+dy[i]
            nx=x+dx[i]
            if 0<=ny<m and 0<=nx<n:
                if not visited[ny][nx] and board[ny][nx]==color:
                    cnt+=dfs(ny,nx,color)
        return cnt
    for i in range(m):
        for j in range(n):
            if not visited[i][j]:
                cnt=dfs(i,j,board[i][j])
                if board[i][j]=='W':
                    t_white+=cnt*cnt
                else:
                    t_blue+=cnt*cnt
    return t_white,t_blue
mii=lambda:sys.stdin.readline().rstrip()
n,m=map(int,mii().split())
board=[mii() for _ in range(m)]
white,blue=calculate_power(n,m,board)
print(white,blue)

File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.cc
#include <iostream>
#include <vector>
using namespace std;
int dy[4] = {-1, 1, 0, 0};
int dx[4] = {0, 0, -1, 1};
int dfs(vector<vector<bool>> &visited, vector<string> &board, int y, int x, char color, int m, int n) {
    int ny, nx, count = 1;
    visited[y][x] = true;
    for (int i = 0; i < 4; i++) {
        ny = y + dy[i];
        nx = x + dx[i];
        if (0 <= ny && ny < m && 0 <= nx && nx < n) {
            if (!visited[ny][nx] && board[ny][nx] == color) {
                count += dfs(visited, board, ny, nx, color, m, n);
            }
        }
    }
    return count;
}
vector<int> calculate_power(int n, int m, vector<string> &board) {
    vector<vector<bool>> visited(m, vector<bool>(n, false));
    int cnt;
    vector<int> t(2, 0);
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (!visited[i][j]) {
                cnt = dfs(visited, board, i, j, board[i][j], m, n);
                if (board[i][j] == 'W')
                    t[0] += cnt * cnt;
                else
                    t[1] += cnt * cnt;
            }
        }
    }
    return t;
}
int main(void) {
    int n, m;
    vector<int> result(2);
    cin >> n >> m;
    vector<string> board(m);
    for (int i = 0; i < m; i++)
        cin >> board[i];
    result = calculate_power(n, m, board);
    cout << result[0] << ' ' << result[1];
    return 0;
}